### DISCUSSION TITLE
Whisper Models are Poisoned?

### DISCUSSION THREAD

**Discussion by caoccao** [openai/whisper#1783] [General]
I was glad to hear that OpenAI released the Whisper v3 and tried right after it went online. However, I found the models are slightly poisoned.

I usually batch transcribe some talk shows in English or Mandarin on MBP M2 Max. Some of them are 4-5 hours long so that the official [whisper](https://github.com/openai/whisper) is too slow. So I use [whisper.cpp](https://github.com/ggerganov/whisper.cpp) which is insanely fast (10+ times).

After upgrading to large-v3, I found the srt files are full of the following ads. As you can see, the ads repeats every 20 seconds forever. The ads literally mean to subscribe an online channel.

<img width="677" alt="ads-from-large-v3" src="https://github.com/openai/whisper/assets/17514279/598324e8-e1ba-4b5b-9157-13fd119d113b">

I struggled for quite some time and thought that might be whisper.cpp's problem. So I turned to the official whisper. Unfortunately, it's the same except it's 10+x slower.

I had to revert back to large-v2. The result is much better, but still full of the ads at the beginning of each talk show as follows.

<img width="1032" alt="ads-from-large-v2" src="https://github.com/openai/whisper/assets/17514279/bb73f109-4344-4df3-bb5e-b12447d31f3b">

The pattern of the ads seemed to be clear to me. Many subtitle editors like to embed ads in the front of the videos. These videos usually play silence or soft music at the beginning. The whisper models learned the silence or soft music are represented by those ads.

Thus, I concluded that the whisper models are poisoned by the training data. Hope OpenAI address and fix this issue.

**Comment by ryanheise** [comment]
This problem exists in both v3 and v2 and has to do with Whisper hallucinating, especially when encountering periods of silence or no speech, and then it can also get stuck in a sort of loop like this. The difference is that the hallucinations probably just occur in different places in v2 and v3, but they still will happen in both cases. If you like whisper.cpp, you can make a feature request to limit this issue by preprocessing the file to cut out the silent or no speech parts first, then re-inserting those parts in the timestamps in postprocessing. There are other projects that already provide this feature, such as stable-ts and others (if you search the discussion board for hallucinations, you can find several more projects that work on improving the hallucination problem.)

**Reply by despairTK** [reply]
I think large-v3 has much more serious problems than large-v2. On the first day I used Whisper's large-v3 to transcribe less than 10 minutes of Japanese audio with repeated sentences.

Then whisper.cpp, which supports large-v3, had the same problem.
https://github.com/ggerganov/whisper.cpp/pull/1444

After faster-whisper supports large-v3, I also tested about a dozen audios, Japanese, Korean, and English, with audio durations ranging from less than 10 minutes to about 1 hour. These tests still have problems with duplicate sentences, incorrect sentence splitting, etc.

 The large-v3 model does not seem to be as good as large-v2, and the biggest improvement I've seen with the large-v3 model is that words with similar pronunciations are selected as the correct one. But the rest is not as good as large-v2. I did these tests without using the translation function.

**Reply by theblackcat102** [reply]
Had the same conclusion here. There's some missing transcription in v3. English wise, I don't see much upside to upgrade it

**Reply by FurkanGozukara** [reply]
very interesting

i also noticed this. v2 v3 hallucinates whereas v1 not

**Comment by caoccao** [comment]
> This problem exists in both v3 and v2 and has to do with Whisper hallucinating, especially when encountering periods of silence or no speech, and then it can also get stuck in a sort of loop like this. The difference is that the hallucinations probably just occur in different places in v2 and v3, but they still will happen in both cases. If you like whisper.cpp, you can make a feature request to limit this issue by preprocessing the file to cut out the silent or no speech parts first, then re-inserting those parts in the timestamps in postprocessing. There are other projects that already provide this feature, such as stable-ts and others (if you search the discussion board for hallucinations, you can find several more projects that work on improving the hallucination problem.)

Thank you for the info. I'll investigate that. In this case, I don't quite believe it's the hallucination. To me, it's quite obvious that the training data is broken.

**Reply by ryanheise** [reply]
Do you know if there was a silent or no-speech region at or prior to when it started misbehaving?

**Reply by caoccao** [reply]
> Do you know if there was a silent or no-speech region at or prior to when it started misbehaving?

Does it consistently output the same ads in that case? A simple grep over the training data can tell. That's the purpose of my creating this discussion.

**Reply by ryanheise** [reply]
I don't have the training data, so I can't do anything about that. Hallucinations are present in all models, one could even say that they are inherent in this type of model,  but there are things outside of the models themselves that we can tune to lean towards correct output. There is a section of python code within whisper that is intended to detect repetitions and then increase the temperature to kick itself out of the loop, which has been there since v1, but this section of code does have a couple of ways in which it can fail. One of those ways is the silence or no speech regions, hence my last question, but if you're not interested in investigating that, that's ok.

**Reply by caoccao** [reply]
My purpose is to inspire OpenAI folks to sanitize the training data, not to discuss with you on the hallucination. I might jump in another topic discussing the hallucination, but not this one. My avoiding discussing hallucination here doesn't mean I'm interested or not interested in it. Hope we both understand each other and walk away.

**Reply by ryanheise** [reply]
It's worth noting that the Whisper whitepaper aimed to show that excellent results could be achieved by sacrificing data quality in order to scale to a massive training dataset size - that was sort of the hypothesis they were testing, and they appear to have been successful in that. So it appears that a massive training set is hugely beneficial even if the data is not perfect, and it is the allowance for imperfections that actually allows you to reach that massive size.

I would agree that better data sanitisation would lead to even better results. Whisper already does some cheap sanitisation.

But I think it would be feasible to also consider a mildly non-cheap form of sanitation that analyses all 5 million hours of audio data by running a VAD algorithm over it, and then filtering all of the spurious text that occurs in a no speech region. After all, this processing need only be done once before training starts.

(But with the models we have today, we can also get more performance out of them by preprocessing with VAD, or by tuning some of the whisper parameters like `no_speech_threshold` and others.)

**Reply by ILG2021** [reply]
It is very normal, I think whisper use many youtube video for train. So many subtitle in the video has no voice. So when we use whisper in real world and encounter silence voice, it will appear this things. int8 inference will appear more frequently. I solve this problem by changing the inference model to float16 and finetune the model with some noise and silence voices. My model nearly no ads again.

**Comment by glangford** [comment]
Does the training data contain any "negative" examples in an attempt to suppress hallucination? eg. Non-speech sounds of applause, instrumental music, silence, chalk on a blackboard, wind, etc with the desired transcription of these being no text at all.

**Comment by jongwook** [comment]
Hi, would it be possible to share the audio file or a link to the talk show you're transcribing, we're aware that hallucination is quite annoying, but we can try some hyperparameter tuning which might be helpful as @glangford mentioned.

**Reply by caoccao** [reply]
Thank you for the reminder. I uploaded a sample to this [shared folder](https://drive.google.com/drive/folders/1_X9lYuNJqdLSO4OL-Ah5MR94uf3rajg6?usp=sharing). The srt file from large-v2 is also uploaded. The subtitles 1-19 and 1336 are either ads or credits that don't exist in the audio file. Especially subtitle 1 is the overture of the famous song **Hotel California** in guitar. I seriously doubt the training data shares a similar pattern.

**Reply by glangford** [reply]
FYI, a quick comparison of V2 and V3 (using the same settings on the uploaded sample). 

`--patience 2 --beam_size 5 --output_format srt --verbose True --language Mandarin --no_speech_threshold 0.5 --condition_on_previous_text False`

### --model large-v2
[00:30.000 --> 00:33.000] 中文字幕由 Amara.org 社群提供
[01:00.000 --> 01:05.000] 技术与资本联姻,成功与风险共存。
[01:05.000 --> 01:08.000] 用睿智投资未来。
[01:08.000 --> 01:14.000] 百名数字英雄跪据京城,献计献策,共商发展。
[01:14.000 --> 01:18.000] 百家讲坛为您奉上精彩系列节目,
[01:18.000 --> 01:24.000] 2001北京互联网发展论坛,敬请关注。
[01:30.000 --> 01:37.000] 江昭阳
[01:37.000 --> 01:39.000] 搜狐网首席执行官
[01:39.000 --> 01:43.000] 1981年考入清华大学物理系
[01:43.000 --> 01:47.000] 1986年考取李正道奖学金赴美读学
[01:47.000 --> 01:51.000] 7年后获麻省理工学院物理学博士
[01:51.000 --> 01:53.000] 1995年10月回国
[01:53.000 --> 01:57.000] 1998年成功推出搜狐网站
[01:57.000 --> 01:58.000] 同年10月


### --model large-v3
[00:00.000 --> 00:29.980] 优优独播剧场——YoYo Television Series Exclusive
[00:30.000 --> 00:59.980] 请不吝点赞 订阅 转发 打赏支持明镜与点点栏目
[01:00.000 --> 01:02.480] 技术与资本联姻
[01:02.480 --> 01:04.620] 成功与风险共存
[01:04.620 --> 01:07.540] 用睿智投资未来
[01:07.540 --> 01:09.760] 百名数字英雄
[01:09.760 --> 01:10.960] 规矩精诚
[01:10.960 --> 01:12.200] 献计献策
[01:12.200 --> 01:13.560] 共商发展
[01:13.560 --> 01:15.440] 百家讲坛
[01:15.440 --> 01:17.780] 为您奉上精彩系列节目
[01:17.780 --> 01:21.120] 2001北京互联网发展论坛
[01:21.120 --> 01:22.740] 敬请关注
[01:30.000 --> 01:59.980] 请不吝点赞 订阅 转发 打赏支持明镜与点点栏目

**Reply by caoccao** [reply]
Thank @glangford for posting the comparisons.

> large-v2
> [00:30.000 --> 00:33.000] 中文字幕由 Amara.org 社群提供

This is an ad.

> large-v3
> [00:00.000 --> 00:29.980] 优优独播剧场——YoYo Television Series Exclusive
> [00:30.000 --> 00:59.980] 请不吝点赞 订阅 转发 打赏支持明镜与点点栏目
> [01:30.000 --> 01:59.980] 请不吝点赞 订阅 转发 打赏支持明镜与点点栏目

These are ads.

---

As the pattern is quite consistent, I think I will keep using large-v2 and write a script to clean up the ads.

**Reply by ryanheise** [reply]
By the way @caoccao would you mind sharing what transcript should have been if it were 100% accurate? I would like to test some fixes to see if I can get it to give the correct output.

I'll also share my base level Whisper output before making any further changes.

```
$ whisper --model large-v2 --language Mandarin --word_timestamps True filename

[00:00.000 --> 00:00.900] 好
[00:37.940 --> 00:53.260] 中文字幕由 Amara.org 社群提供
[01:00.880 --> 01:04.640] 技術與資本聯姻成功與風險共存
[01:04.640 --> 01:07.580] 用睿智投資未來
[01:08.200 --> 01:11.060] 百名數字英雄匯聚京城
[01:11.060 --> 01:13.620] 獻計獻策 共商發展
[01:14.340 --> 01:17.720] 百家講壇為您奉上精彩系列節目
[01:17.720 --> 01:21.120] 2001北京互聯網發展論壇
[01:21.120 --> 01:22.760] 敬請關注
[01:44.240 --> 01:50.640] 中文字幕由 Amara.org 社群提供
```

```
$ whisper --model large-v3 --language Mandarin --word_timestamps True filename

[00:24.060 --> 00:28.200] 字幕由 Amara.org 社群提供
[00:57.760 --> 00:59.680] 明镜需要您的支持 欢迎订阅明镜
[01:21.440 --> 01:29.900] 明镜需要您的支持 欢迎订阅明镜
[01:58.200 --> 01:59.980] 明镜需要您的支持 欢迎订阅明镜
```

Now this is the v3 model again after some tweaks:

```
互联网
一个最为开放的领域
一个汇聚精英的世界
技术与资本联姻
成功与风险共存
用睿智投资未来
百名数字英雄汇聚京城
献计献策
共商发展
百家讲坛
为您奉上精彩系列节目
2001北京互联网发展论坛
敬请关注
张昭阳
搜户网首席执行官
1981年考入清华大学物理系
1986年考取理政道奖学金赴美留学
七年后获麻省理工学院物理学博士
1995年
10月
回国
1998年
成功推出搜户网站
同年10月
为美国时代周刊
评委
```

**Reply by caoccao** [reply]
@ryanheise , sorry I don't have the 100% correct script. Basically, it's not worth my time correcting it as I have thousands of such audios.

> Now this is the v3 model again after some tweaks

I think this version is a lot better.

**Reply by ryanheise** [reply]
> it's not worth my time correcting it as I have thousands of such audios.

I only need the correct transcript for the one example above, in order to test the fix. Once you have the fix, then your thousands of other audios can benefit from the fix.

If any native Chinese speakers are listening in, would you mind taking a look at the transcript above and pointing out any corrections, please?

**Reply by caoccao** [reply]
I cleaned up the noises from large-v2 and corrected a few lines. I have to admit the result you got from large-v3 is no better than large-v2. The one you got is in simplified Chinese, and the one I got is in traditional Chinese. This is another weird thing that whisper does: simplified or traditional Chinese seems to be in random. You could easily translate them using an online service to verify your result. Thank you.

```
19
00:00:54,000 --> 00:00:56,000
互聯網

20
00:00:56,000 --> 00:00:58,000
一個最為開放的領域

21
00:00:58,000 --> 00:01:00,000
一個匯聚精英的世界

22
00:01:00,000 --> 00:01:03,000
技術與資本聯姻

23
00:01:03,000 --> 00:01:05,000
成功與風險共存

24
00:01:05,000 --> 00:01:07,000
用睿智投資未來

25
00:01:07,000 --> 00:01:10,000
百名數字英雄

26
00:01:10,000 --> 00:01:11,000
匯聚京城

27
00:01:11,000 --> 00:01:12,000
獻計獻策

28
00:01:12,000 --> 00:01:14,000
共商發展

29
00:01:14,000 --> 00:01:18,000
百家講壇為您奉上精彩系列節目

30
00:01:18,000 --> 00:01:21,000
2001北京互聯網發展論壇

31
00:01:21,000 --> 00:01:23,000
敬請關注

34
00:01:28,000 --> 00:01:37,000
張朝陽

35
00:01:37,000 --> 00:01:39,000
搜狐網首席執行官

36
00:01:39,000 --> 00:01:42,000
1981年考入清華大學物理系

37
00:01:42,000 --> 00:01:45,000
1986年考取李政道獎學金

38
00:01:45,000 --> 00:01:47,000
赴美留學

39
00:01:47,000 --> 00:01:51,000
七年後獲麻省理工學院物理學博士

40
00:01:51,000 --> 00:01:53,000
1995年10月回國

41
00:01:53,000 --> 00:01:56,000
1998年成功推出搜狐網站

42
00:01:56,000 --> 00:01:59,000
同年10月為美國時代周刊

43
00:01:59,000 --> 00:02:01,000
評為全球數字英雄
```

**Reply by ryanheise** [reply]
Thank you for supplying that.

> I have to admit the result you got from large-v3 is no better than large-v2

Do you mean without the fix? Clearly the v3 output without the fix is "worse" than v2 without the fix.
But with the fix, I only showed the fixed v3 output and not the fixed v2 output, so I'm not sure what you might be comparing it to.

> This is another weird thing that whisper does: simplified or traditional Chinese seems to be in random.

Yes it may initially be random, but once it starts transcribing, the next window will likely continue in the style of the previous windows. Since there are no previous windows before the first window, the very first window will probably have some chance to it. But you can insert some context before the first window to influence the style of the first window like this: https://github.com/openai/whisper/discussions/277#discussioncomment-3832154

**Reply by caoccao** [reply]
> Do you mean without the fix? Clearly the v3 output without the fix is "worse" than v2 without the fix.
> But with the fix, I only showed the fixed v3 output and not the fixed v2 output, so I'm not sure what you might be comparing it to.

There are a few obvious errors or typos in your revised v3 output. My rough estimate is the error or typo rate between v2 and v3 is subtle, which means human would probably spend the same amount of time correcting them. So I think the quality is the same as well. As I mentioned in my early reply, I could write a script easily clean up most of the ads automatically from v2, so I'm living with v2 for now.

Regarding the simplified or traditional Chinese topic, thank you for the info. Actually, I kind of enjoy whisper's random choice as I'm good at reading / writing both.

**Reply by ryanheise** [reply]
OK, I won't be able to do anything about the typos actually, so I should have asked for the less burdensome task of correcting only the hallucinations, if any. I have run my v3 output through a simplified to traditional converter and then did a diff against your corrected output and noticed a few differences in spelling as you suggested, although no text was outright added or removed, so I think this means that at least the hallucinations were removed successfully. I'll try to sort this out in the code. And of course you can alternatively try one of the projects referenced earlier that solve the issue through VAD preprocessing.

**Reply by ryanheise** [reply]
I've created PR #1838 which skips over any silence before a detected hallucination that is longer than `--hallucination_silence_threshold DURATION` in seconds (I used 2 seconds). On the PR page I've included some sample output of your example with a threshold of 2 seconds, and included some debug output to also print out whenever a hallucination was detected. The PR only skips over parts where Whisper didn't detect any speech, although Whisper can still fail to detect speech when there was some, and this PR doesn't address that.

I can confirm this works more reliably with v2 than v3 on your example. For it to work with v3, it has to skip precisely the right amount of silence at the start for it to successfully pick up the start of the first utterance, but v3 has harder picking up the start of that speech. This might be because the speech starting at around 54 seconds is very soft/inaudible at the start. But v2 is more successful at picking up the first utterance.

**Reply by dgoryeo** [reply]
@ryanheise , I just noticed that your [ryanheise:fix-hallucinations](https://github.com/ryanheise/whisper/tree/fix-hallucinations) was commited to openai:main. Great job!!!   Am I right to understand that both options `--hallucination_silence_threshold` , and `--clip_timestamps` are now avilable?  Any tips on optimal values?

**Reply by ryanheise** [reply]
They are both part of the same PR, so you can use both. I used `--hallucination_silence_threshold 2` which will make Whisper vigilant against hallucinations after encountering at least 2 seconds of silence. What is the optimal threshold? There is probably no point in making Whisper vigilant against hallucinations after 0.3 seconds of silence, because that's probably the speaker just taking a silent breath between words, and you don't want to waste processing time on it. On the other hand if you set the bar too high at 15 seconds, then Whisper will not bother to notice hallucinations occurring after 8 seconds of silence. So I wouldn't go lower than 2 seconds, and I wouldn't go higher than 8 seconds. Somewhere in that range there might be an optimal value.

As for `--clip_timestamps`, this option is there for people who want to use some external tool to find where there is actual speech and where there is silence, and then pass those timestamps into Whisper, instead of asking Whisper to detect it via the previous option. A separate tool that can do this would be something like Silero VAD. That tool will take its own threshold parameters to specify how long a silent period needs to be to register. I suggest not setting it too low, because as above, there is no point in vigilantly skipping over 0.3 seconds of silence and you won't want to waste processing time on it.

**Comment by turicas** [comment]
I have a similar problem, but with audios with no sounds. More details at: https://github.com/openai/whisper/discussions/1606#discussioncomment-7541656

**Reply by caoccao** [reply]
Exactly, the problem seems not to be related to the hallucination, but the broken training data.

**Reply by ryanheise** [reply]
It's not that it is either one or the other, hallucinations can come about on "no sounds" because the training data had some kind of spurious subtitles during no speech (e.g. credits/copyrights/ads). But they are still called hallucinations.

**Reply by turicas** [reply]
One way to improve the results is to pass a VAD filter on the audio file before sending to the model (`model.transcribe(..., vad_filter=True)`).

**Reply by ryanheise** [reply]
There isn't a `vad_filter` option in Whisper itself, although there are many other unofficial projects that do this, perhaps the one you are referring to is [faster-whisper](https://github.com/guillaumekln/faster-whisper).

**Comment by turicas** [comment]
> There isn't a `vad_filter` option in Whisper itself, although there are many other unofficial projects that do this, perhaps the one you are referring to is [faster-whisper](https://github.com/guillaumekln/faster-whisper).

Yes, I'm using faster-whisper! Sorry for the confusion. [faster-whisper uses Silero VAD filter](https://github.com/guillaumekln/faster-whisper/blob/master/faster_whisper/vad.py).

**Comment by lennondotw** [comment]
Same issue here, using official ChatGPT iOS App, 4o model and Advanced Voice Mode.

![image](https://github.com/user-attachments/assets/6e3bd7cd-3554-426f-b0b8-6338ba6f714a)

**Comment by HIRSHDZ327** [comment]
I tried many times to get a good one to use, the command line:

whisper.exe "Your Document" --model large-v3-turbo --device cuda --language en --task transcribe --word_timestamps True --temperature 0.2   --no_speech_threshold 0.3 --fp16 True --word_timestamps True --condition_on_previous_text False --max_line_width 100 --output_format srt --suppress_tokens -1 --initial_prompt "This is about xxxx(what your vedio or audio actually is about.)" 

I'd use v3-turbo model,cause it was just fast enough,and the accuracy didn't fall back that much,just in my perspective,probably 5%-15%?
but A looooooooot faster.
use cuda,designated the language,the type of task,and what's next...  Well, there's  just too much to type. I'd just leave the rest of the explanation to ChatGPT or Deepseek.（Meaning，you can google it or Ask one of those AIs）

Well ,that "max_line_width"parameter is more decided by  your device's screen resolution,and how big a character actually is on your screen ,let AI do the calculation,but normally 100 should be fine,if you think it might be too many,40/60/80,these are usually my choices.

**Comment by lanyangyangyeyeye** [comment]
I have same issue when  I  finetuning whisper-small . after training ,I found the cer is 250%,which is strange.but when I use the trained model to transcript an audio,it works.so I'm so confused.Then I compare the code ,I found that when I change `forced_decoder_ids = processor.get_decoder_prompt_ids(language="zh", task="transcribe")`into `forced_decoder_ids = processor.get_decoder_prompt_ids(language="chinese", task="transcribe")`.The hallucinations is gone .So ,If you are using whisper model to transcript an audio ,you can try this way to figure the problem.

**Comment by zhaoniubi** [comment]
![Uploading image.png…]()

**Comment by LOVECHEN** [comment]
still have this issue

**Comment by LiuJunheng** [comment]
还没有修复，太坑人了
